src/AlertBar/index.js

```diff
@@ -0,0 +1,146 @@
+        if (
+            prevState.visible === this.state.visible &&
+            prevState.hidden === this.state.hidden
+        ) {
+            this.init()
+        }
+    }
+
+    componentWillUnmount() {
+        clearTimeout(this.displayTimeout)
+        clearTimeout(this.onHiddenTimeout)
+    }
+
+    init() {
+        this.startTime = Date.now()
+        this.timeRemaining = this.props.duration
+        this.startDisplayTimeout()
+        this.show()
+    }
+
+    startDisplayTimeout = () => {
+        if (this.shouldAutoHide()) {
+            this.displayTimeout = setTimeout(() => {
+                this.hide()
+            }, this.timeRemaining)
+        }
+    }
+
+    stopDisplayTimeOut = () => {
+        if (this.shouldAutoHide()) {
+            const elapsedTime = Date.now() - this.startTime
+            this.timeRemaining = this.timeRemaining - elapsedTime
+            clearTimeout(this.displayTimeout)
+        }
+    }
+
+    hide = () => {
+        clearTimeout(this.displayTimeout)
+        this.setState({ visible: false })
+
+        this.onHiddenTimeout = setTimeout(() => {
+            this.setState({ hidden: true })
+            this.props.onHidden && this.props.onHidden()
+        }, ANIMATION_TIME)
+    }
+
+    show() {
+        requestAnimationFrame(() => {
+            this.setState({ visible: true })
+        })
+    }
+
+    shouldAutoHide() {
+        const { permanent, warning, critical } = this.props
+        return !(permanent || warning || critical)
+    }
+
+    render() {
+        const {
+            className,
+            children,
+            success,
+            warning,
+            critical,
+            icon,
+            actions,
+        } = this.props
+        const { visible, hidden } = this.state
+
+        if (hidden) {
+            return null
+        }
+
+        const info = !critical && !success && !warning
+        const iconProps = { icon, critical, success, warning }
+
+        return (
+            <div
+                className={cx(className, {
+                    info,
+                    success,
+                    warning,
+                    critical,
+                    visible,
+                })}
+                onMouseEnter={this.stopDisplayTimeOut}
+                onMouseLeave={this.startDisplayTimeout}
+            >
+                <Icon {...iconProps} />
+                <Message>{children}</Message>
+                <Actions actions={actions} hide={this.hide} />
+                <Dismiss onClick={this.hide} />
+
+                <style jsx>{styles}</style>
+            </div>
+        )
+    }
+}
+
+const variantPropType = mutuallyExclusive(
+    ['success', 'warning', 'critical'],
+    propTypes.bool
+)
+
+AlertBar.propTypes = {
+    className: propTypes.string,
+    children: propTypes.string.isRequired,
+    success: variantPropType,
+    warning: variantPropType,
+    critical: variantPropType,
+    icon: iconPropType,
```

Comment:
This is ambiguous, I'd prefer to have a 'noIcon` prop, but it's not too important. I'd reduce the "complexity" in the `Icon` component as well (complexity in quotes because it's not complex at all)

Reply:
I see your point. However, @varl and I discussed this before. We are introducing some complexity here, but we thought it provides a very convenient API (see the Icons story for an illustration).